src/components/page-header/change-type-modal.tsx

```diff
@@ -0,0 +1,371 @@
+import { LineId, MiscNodeId, StnId, Theme } from '../../constants/constants';
+import { LinePathType, LineStyleType } from '../../constants/lines';
+import { StationType } from '../../constants/stations';
+import { useRootDispatch, useRootSelector } from '../../redux';
+import { saveGraph } from '../../redux/param/param-slice';
+import { openPaletteAppClip, setRefreshEdges, setRefreshNodes } from '../../redux/runtime/runtime-slice';
+import { findThemes } from '../../util/graph';
+import {
+    changeLinePathTypeInBatch,
+    changeLinesColorInBatch,
+    changeLineStyleTypeInBatch,
+    changeNodesColorInBatch,
+    changeStationsTypeInBatch,
+} from '../../util/change-types';
+import { linePaths, lineStyles } from '../svgs/lines/lines';
+import stations from '../svgs/stations/stations';
+import ThemeButton from '../panels/theme-button';
+
+export type FilterType = 'station' | 'misc-node' | 'line';
+
+interface ChangeTypeInfo {
+    id: string;
+    title: string;
+    onClose: () => void;
+    field: RmgFieldsField[];
+}
+
+interface ChangeTypeTheme {
+    id: string;
+    theme: Theme;
+    value: string;
+}
+
+export const ChangeTypeModal = (props: {
+    isOpen: boolean;
+    onClose: () => void;
+    isSelect: boolean;
+    filter?: FilterType[];
+}) => {
+    const { isOpen, onClose, isSelect, filter } = props;
+    const { t } = useTranslation();
+    const dispatch = useRootDispatch();
+    const { selected, theme, paletteAppClip: output } = useRootSelector(state => state.runtime);
+
+    const hardRefresh = React.useCallback(() => {
+        dispatch(setRefreshNodes());
+        dispatch(setRefreshEdges());
+        dispatch(saveGraph(graph.current.export()));
+    }, [dispatch, setRefreshNodes, setRefreshEdges, saveGraph]);
+    const graph = React.useRef(window.graph);
+
+    const availableLinePathOptions = {
+        any: t('header.settings.procedures.changeType.any'),
+        ...(Object.fromEntries(
+            Object.entries(linePaths).map(([key, val]) => [key, t(val.metadata.displayName).toString()])
+        ) as { [k in LinePathType]: string }),
+    };
+    const availableLineStyleOptions = {
+        any: t('header.settings.procedures.changeType.any'),
+        ...(Object.fromEntries(
+            Object.entries(lineStyles).map(([key, val]) => [key, t(val.metadata.displayName).toString()])
+        ) as { [k in LineStyleType]: string }),
+    };
+    const availableStationOptions = {
+        any: t('header.settings.procedures.changeType.any'),
+        ...(Object.fromEntries(
+            Object.entries(stations).map(([key, val]) => [key, t(val.metadata.displayName).toString()])
+        ) as { [k in StationType]: string }),
+    };
+
+    const [isStationTypeSwitch, setIsStationTypeSwitch] = React.useState(false);
+    const [currentStationType, setCurrentStationType] = React.useState<StationType | 'any'>('any');
+    const [newStationType, setNewStationType] = React.useState<StationType>(StationType.ShmetroBasic);
```

Comment:
This type hint is redundant as ts could infer it automatically.

Reply:
It seems that I need to add `as ...` when I call the functions like `changeStationsTypeInBatch` if I delete the type hint here.

<img width="705" alt="image" src="https://github.com/railmapgen/rmp/assets/59787082/e0a0fe46-441a-41aa-a41f-2e9a366c0d66">